src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts

```diff
@@ -0,0 +1,296 @@
+    }
+  }
+
+  /**
+   * Updates model value with the selected value
+   * @param event The value to set.
+   */
+  onSelectItem(event: NgbTypeaheadSelectItemEvent) {
+    this.updateModel(event.item);
+    this.cdr.detectChanges();
+  }
+
+  onBlur(event: Event) {
+    this.dispatchUpdate(this.currentValue);
+    this.cdr.detectChanges();
+  }
+
+  updateModel(updateValue) {
+    let newValue = updateValue.display;
+    if (this.isSponsorInputType) {
+      if (updateValue instanceof VocabularyEntry) {
+        newValue = AUTOCOMPLETE_COMPLEX_PREFIX + SEPARATOR + updateValue.value;
+      } else {
+        // special autocomplete sponsor input
+        newValue = this.composeSponsorInput(updateValue);
+      }
+    }
+    this.dispatchUpdate(newValue);
+  }
+
+  /**
+   * Emits a change event and updates model value.
+   * @param newValue
+   */
+  dispatchUpdate(newValue: any) {
+    this.model.value = newValue;
+    this.change.emit(newValue);
+  }
+
+  /**
+   * Sets the current value with the given value.
+   * @param value given value.
+   * @param init is initial value or not.
+   */
+  public setCurrentValue(value: any, init = false) {
+    let result: string;
+    if (init) {
+      this.getInitValueFromModel()
+        .subscribe((formValue: FormFieldMetadataValueObject) => {
+          this.currentValue = formValue;
+          this.cdr.detectChanges();
+        });
+    } else {
+      if (isEmpty(value)) {
+        result = '';
+      } else {
+        result = value.value;
+      }
+
+      this.currentValue = result;
+      this.cdr.detectChanges();
+    }
+  }
+
+  formatter = (x: { display: string }) => {
+    return x.display;
+  }
+
+  suggestionFormatter = (suggestion: TemplateRef<any>) => {
+    if (this.isSponsorInputType) {
+      let formattedValue = '';
+      // normal MetadataValue suggestion
+      if (suggestion instanceof VocabularyEntry) {
+        formattedValue = suggestion.value;
+        if (formattedValue.startsWith(AUTOCOMPLETE_COMPLEX_PREFIX)) {
+          formattedValue = DynamicAutocompleteService.removeAutocompletePrefix(suggestion);
+        }
+        const complexInputList = formattedValue.split(SEPARATOR);
+        return DynamicAutocompleteService.pretifyFundingSuggestion(complexInputList[1], complexInputList[2]);
+      } else if (suggestion instanceof ExternalSourceEntry) {
+        // suggestion from the openAIRE
+        const fundingProjectCode = this.getProjectCodeFromId(suggestion.id);
+        const fundingName = suggestion.metadata['project.funder.name'][0].value;
+        return DynamicAutocompleteService.pretifyFundingSuggestion(fundingProjectCode, fundingName);
+      }
+    }
+    // @ts-ignore
+    return suggestion.display;
+  }
+
+  /**
+   * Converts a text values stream from the `<input>` element to the array stream of the items
+   * and display them in the typeahead popup.
+   */
+  search = (text$: Observable<string>) =>
+    text$.pipe(
+      debounceTime(300),
```

Comment:
?

Reply:
I don't know I copied it from the _dynamic-tag.component_ (DSpace class). But I think 300ms as debounce time (time between typing a characters) is good option.